[Doc] Improve type.rst readability (#1618)
taichi-dev/taichi: docs/type.rst

Type system
===========

> Taichi supports many data types. The type name is recognized as
> a *prefix character* + a *digital number*.

Reviewer: ```suggestion
Taichi supports common numerical data types. Each type is denoted as
a character indicating its *category* and a number of *precision bits*, e.g., ``i32`` and ``f64``.
```

WDYT? If you think it's clearer please change the rest accordingly.

> The *prefix character* can be one of:

Reviewer: ```suggestion
The *category* can be one of:
```

    (OK: supported, EXT: require extension, N/A: not available)

.. note::

    Boolean types are represented using ``ti.i32``.

Type promotion
--------------

Binary operations on different types will give you a promoted type, following the C programming language convention, e.g.:

- ``i32 + f32 = f32`` (integer + float = float)
- ``i32 + i64 = i64`` (less-bits + more-bits = more-bits)

Basically it will try to choose the more precise type to contain the result value.

.. _default_precisions:

Default precisions
------------------

By default, all numerical literals have 32-bit precisions.
For example, ``42`` has type ``ti.i32`` and ``3.14`` has type ``ti.f32``.

Default integer and float-point precisions (``default_ip`` and ``default_fp``) can be specified when initializing Taichi:

.. code-block:: python

    ti.init(default_fp=ti.f32)
    ti.init(default_fp=ti.f64)

    ti.init(default_ip=ti.i32)
    ti.init(default_ip=ti.i64)

Also note that you may use ``float`` or ``int`` in type definitions as aliases
for default precisions, e.g.:

.. code-block:: python

    ti.init(default_ip=ti.i64, default_fp=ti.f32)

>     x = ti.var(float, 5)
>     y = ti.var(int, 5)
>     # is equivalent to:
>     x = ti.var(ti.f32, 5)
>     y = ti.var(ti.i64, 5)

Reviewer: ```suggestion
    x = ti.field(dtype=float, shape=5)
    y = ti.field(dtype=int, shape=5)
    # is equivalent to:
    x = ti.field(dtype=ti.f32, shape=5)
    y = ti.field(dtype=ti.i64, shape=5)
```

Author: Thanks, let's only do `ti.var` -> `ti.field` after `rullec-field`, wdyt?

Reviewer: OK, I think that's a good reason to stick to `ti.var` for now.

    def func(a: float) -> int:
        ...

    # is equivalent to:
    def func(a: ti.f32) -> ti.i64:
        ...

Type casts
----------

Implicit casts
**************

> WARNING: The type of a variable is **determinated on it's initialization**.

Reviewer: Maybe use a `warning` block?
```
.. warning::
  ....
```

Author: ```suggestion
.. warning::

  The type of a variable is **determinated on it's initialization**.
```

> When a *wide* variable is assigned with a *narrow* type, it will be
> implicitly promoted to the *wide* type and no warning will be raised:

Reviewer: ```suggestion
When a *low-precision* variable is assigned to a *high-precision* variable, it will be
implicitly promoted to the *wide* type and no warning will be raised:
```

.. code-block:: python

    a = 1.7
    a = 1
    print(a)  # 1.0

> When a *narrow* variable is assigned with a *wide* type, it will be
> implicitly casted into the *narrow* type and Taichi will raise a warning:

Reviewer: ```suggestion
When a *high-precision* variable is assigned to a *low-precision* type, it will be
implicitly down-cast into the *low-precision* type and Taichi will raise a warning:
```

.. code-block:: python

    a = 1
    a = 1.7
    print(a)  # 1

Explicit casts
**************

You may use ``ti.cast`` to explicitly cast scalar values between different types:

.. code-block:: python

    a = 1.7
    b = ti.cast(a, ti.i32)  # 1
    c = ti.cast(b, ti.f32)  # 1.0

Equivalently, use ``int()`` and ``float()`` to convert values to float-point or
integer types of default precisions:

.. code-block:: python

    a = 1.7
    b = int(a)    # 1
    c = float(a)  # 1.0

> Casting vector / matrix elements
> ********************************

Reviewer: ```suggestion
Casting vectors and matrices
****************************
```

Type casts applied to vectors/matrices are element-wise:

.. code-block:: python

    u = ti.Vector([2.3, 4.7])
    v = int(u)              # ti.Vector([2, 4])

>     # equivalent to:

Reviewer: ```suggestion
    # If you are using ti.i32 as default_ip, this is equivalent to:
```

    v = ti.cast(u, ti.i32)  # ti.Vector([2, 4])

Bit casting
***********

Use ``ti.bit_cast`` to bit-cast a value into another data type. The underlying bits will be preserved in this cast.

> The new type must have the same width as the the old type.

Author: ```suggestion
The new type must have the same width as the the old type.
or example, bit-casting ``i32`` to ``f64`` is not allowed. Use this operation with caution.
```

    a = 1.7
    b = int(a)    # 1
    c = float(a)  # 1.0

Casting vector / matrix elements
********************************

Type casts applied to vectors/matrices are element-wise:

.. code-block:: python

    u = ti.Vector([2.3, 4.7])
    v = int(u)              # ti.Vector([2, 4])
    # equivalent to:
    v = ti.cast(u, ti.i32)  # ti.Vector([2, 4])

Bit casting
***********

Use ``ti.bit_cast`` to bit-cast a value into another data type. The underlying bits will be preserved in this cast.
The new type must have the same width as the the old type.

> For example, bit-casting ``i32`` to ``f64`` is not allowed. Use this operation with caution.

Reviewer: IIRC this is still not allowed?

.. code-block:: python

    a = 1.7
    a = 1
    print(a)  # 1.0

When a *high-precision* variable is assigned to a *low-precision* type, it will be
implicitly down-cast into the *low-precision* type and Taichi will raise a warning:

.. code-block:: python

    a = 1
    a = 1.7
    print(a)  # 1

Explicit casts
**************

You may use ``ti.cast`` to explicitly cast scalar values between different types:

.. code-block:: python

    a = 1.7
    b = ti.cast(a, ti.i32)  # 1
    c = ti.cast(b, ti.f32)  # 1.0

Equivalently, use ``int()`` and ``float()`` to convert values to float-point or
integer types of default precisions:

.. code-block:: python

    a = 1.7
    b = int(a)    # 1
    c = float(a)  # 1.0

Casting vectors and matrices
****************************

Type casts applied to vectors/matrices are element-wise:

.. code-block:: python

    u = ti.Vector([2.3, 4.7])
    v = int(u)              # ti.Vector([2, 4])
    # If you are using ti.i32 as default_ip, this is equivalent to:
    v = ti.cast(u, ti.i32)  # ti.Vector([2, 4])

Bit casting
***********

Use ``ti.bit_cast`` to bit-cast a value into another data type. The underlying bits will be preserved in this cast.

> The new type must have the same width as the the old type.

Author: ```suggestion
The new type must have the same width as the the old type.
For example, bit-casting ``i32`` to ``f64`` is not allowed. Use this operation with caution.
```

implicitly down-cast into the *low-precision* type and Taichi will raise a warning:

.. code-block:: python

    a = 1
    a = 1.7
    print(a)  # 1

Explicit casts
**************

You may use ``ti.cast`` to explicitly cast scalar values between different types:

.. code-block:: python

    a = 1.7
    b = ti.cast(a, ti.i32)  # 1
    c = ti.cast(b, ti.f32)  # 1.0

Equivalently, use ``int()`` and ``float()`` to convert values to float-point or
integer types of default precisions:

.. code-block:: python

    a = 1.7
    b = int(a)    # 1
    c = float(a)  # 1.0

Casting vectors and matrices
****************************

Type casts applied to vectors/matrices are element-wise:

.. code-block:: python

    u = ti.Vector([2.3, 4.7])
    v = int(u)              # ti.Vector([2, 4])
    # If you are using ti.i32 as default_ip, this is equivalent to:
    v = ti.cast(u, ti.i32)  # ti.Vector([2, 4])

Bit casting
***********

Use ``ti.bit_cast`` to bit-cast a value into another data type. The underlying bits will be preserved in this cast.
The new type must have the same width as the the old type.
For example, bit-casting ``i32`` to ``f64`` is not allowed. Use this operation with caution.

> or example, bit-casting ``i32`` to ``f64`` is not allowed. Use this operation with caution.

Author: ```suggestion
```
